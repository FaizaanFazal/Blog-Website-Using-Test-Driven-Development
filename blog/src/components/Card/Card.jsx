import '../../styles/Card.scss';
import { Link } from 'react-router-dom';
import images from '../../utils/images';

export default function Card({ blogItemData, showContent, isCardSm }) {
  // todo get id of author from blog and fetch his image url and name from users

  return (
    <Link to={`/blogdetails/${blogItemData?.slug}`} data-testid="slugLink">
      <div data-testid="artice" className={`card ${isCardSm ? 'card-sm' : ''}`}>
        <img data-testid="Image" src={`${blogItemData?.imageSrc}`} alt={`${blogItemData?.imageAlt}`} className="card-overlay-img object-fit-cover" />
        <div className="card-content">
          <div className="card-body">
            <h3 className="title title-base">
              {blogItemData?.title}
            </h3>
            {
              showContent && <p className="text text-lg">{blogItemData?.content }</p>
            }
          </div>
          <div className="card-footer card-footer-light flex justify-between items-center">
            <div className="writer-info grid">
              <div className="info-avatar">
                <img src={`${blogItemData?.authorImageSrc}`} alt={`${blogItemData?.authorImageAlt}`} className="object-fit-cover" />
              </div>
              <div className="info-intro">
                <p className="intro-name text-base">{ blogItemData?.author }</p>
                <div className="intro-verify flex items-center">
                  <img src={`${images.verify_icon_filled}`} alt="icon" className="verify-icon" />
                  <span className="verify-status text-sm text">Verified writer</span>
                </div>
              </div>
            </div>
            <div className="date-info text text-base">{ blogItemData?.date}</div>
          </div>
        </div>
      </div>
    </Link>
  );
}
