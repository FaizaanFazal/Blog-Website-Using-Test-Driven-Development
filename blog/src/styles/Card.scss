@import "../App";

.card-grid {
  grid-template-columns: repeat(2, 1fr);
  height: 460px;

  .card-grid-text {
    padding: 32px;

    @include breakpoint-down(lg) {
      padding: 24px;
    }
  }

  .card-grid-img {
    @include breakpoint-down(lg) {
      height: 350px;
    }

    @include breakpoint-down(sm) {
      height: 280px;
    }

    @include breakpoint-down(xs) {
      height: 200px;
    }
  }


  @include breakpoint-down(lg) {
    height: auto;
    grid-template-columns: repeat(1, 1fr);
  }
}

.card {
  position: relative;
  border-radius: 16px;
  box-shadow: 0 20px 50px 0 rgb(15 54 44 / 15%);
  transition: $default-transition;
  height: 600px;

  @include breakpoint-down(lg) {
    height: 450px;
  }

  @include breakpoint-down(lg) {
    height: 420px !important;
  }

  @include breakpoint-down(lg) {
    height: 380px !important;
  }

  &-overlay-img {
    border-radius: 16px;
  }

  &::after {
    content: "";
    position: absolute;
    inset: 0;
    background: linear-gradient(180deg, rgb(24 59 86 / 0%) 0%, rgb(22 49 70 / 45%) 45.38%, #152532 100%);
    z-index: 0;
    border-radius: 16px;
    transition: $default-transition;
  }

  .badge {
    position: absolute;
    top: 32px;
    left: 26px;
    z-index: 1;

    @include breakpoint-down(lg) {
      left: 16px;
      top: 16px;
    }
  }

  &-content {
    position: absolute;
    left: 0;
    bottom: 0;
    width: 100%;
    padding: 0 48px 42px;
    z-index: 1;

    @include breakpoint-down(lg) {
      padding: 0 32px 24px;
    }
  }

  &-body {
    .title {
      color: $color-white;
    }

    .text {
      color: $color-white;
    }
  }

  &-footer {
    margin-top: 24px;
    background-color: transparent !important;

    .writer-info {
      grid-template-columns: 40px auto;
      column-gap: 4px;

      .info-avatar {
        width: 40px;
        height: 40px;
        border-radius: 100%;
        overflow: hidden;
      }

      .intro-name {
        color: $color-blue;
        margin-bottom: 4px;
        font-weight: 600;
      }

      .intro-verify {
        column-gap: 8px;
      }

      .verify-icon {
        width: 16px;
      }

      .verify-status {
        font-style: italic;
        font-weight: 300;
      }
    }

    .date-info {
      font-weight: 600;
      color: $color-blue-dark;
    }

    &-light {
      .writer-info {
        .intro-name,
        .verify-status {
          color: $color-white;
        }
      }
    }

    @include breakpoint-down(lg) {
      margin-top: 20px;
    }
  }

  &:hover {
    transform: translateY(-8px);
  }

  &.card-sm {
    .title-base {
      font-size: 24px;
    }

    .card-content {
      padding: 36px 28px;
    }
  }
}
.likebtncont{
  display: flex;

}
.iconlike{
  margin: 15px;
  width: 50px !important;
  height: 50px !important;
}
