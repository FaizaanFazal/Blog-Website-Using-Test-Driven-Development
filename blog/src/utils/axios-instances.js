// import axios from 'axios';

// export const axiosInstance = axios.create({
//   baseURL: 'http://localhost:8000/',
// });
import axios from 'axios';

const placeholderApi = axios.create({
  baseURL: 'http://localhost:8000',
});

export default placeholderApi;
