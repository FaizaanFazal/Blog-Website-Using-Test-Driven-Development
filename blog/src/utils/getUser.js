export const getUser = () => {
    const userString = localStorage.getItem('user');
    if (userString) {
      const user = JSON.parse(userString);
      return user;
    } else {
      return null;
    }
  };