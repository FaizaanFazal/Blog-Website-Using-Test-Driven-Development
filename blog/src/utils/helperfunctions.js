import { fireEvent } from '@testing-library/react';
import images from './images';

export const isProperImageURL = (url) => {
  /* eslint-disable-next-line no-useless-escape */
  const regex = /^(https?|ftp):\/\/(([a-z\d]([a-z\d-]*[a-z\d])?\.)+[a-z]{2,}|localhost)(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(\#[-a-z\d_]*)?$/i;
  return regex.test(url);
};
export const isProperEmail = (email) => {
  const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
};
export const isProperPass = (pass) => {
  const passregex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/;
  return passregex.test(pass);
};
export function backspace(element) {
  let actuallyTyped = element.value;

  const backspaceKey = {
    key: 'Backspace',
    code: 8,
    inputType: 'deleteContentBackward',
  };

  const sharedEventConfig = {
    key: backspaceKey.key,
    charCode: backspaceKey.code,
    keyCode: backspaceKey.code,
    which: backspaceKey.code,
    modifier: backspaceKey.modifier,
  };
  const downEvent = fireEvent.keyDown(element, sharedEventConfig);

  if (downEvent) {
    actuallyTyped = actuallyTyped.slice(0, -1);

    fireEvent.input(element, {
      target: { value: actuallyTyped },
      inputType: backspaceKey.inputType,
      bubbles: true,
      cancelable: true,
    });
  }

  fireEvent.keyUp(element, sharedEventConfig);
}

export const small = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    verified: true,
    slug: 'post1',
    content: "test Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.",
  },
];
export const greater = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: ' Elena' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    slug: 'post1',
    verified: true,
    content: "Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n",
  }];
