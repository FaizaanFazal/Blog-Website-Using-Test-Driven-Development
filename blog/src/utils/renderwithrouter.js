import {render} from '@testing-library/react'
import * as userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup,
    ...render(ui, {wrapper: MemoryRouter}),
  }
}