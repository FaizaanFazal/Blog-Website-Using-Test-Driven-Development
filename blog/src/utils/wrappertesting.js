import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

export function renderWithProviders(
  ui,
  {
    // preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = reduxStore,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

renderWithProviders.propTypes = {
  children: PropTypes.node.isRequired,
};
