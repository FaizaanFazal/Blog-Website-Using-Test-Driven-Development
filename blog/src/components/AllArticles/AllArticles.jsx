import React from 'react';
import { useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import Card from '../Card/Card';

export default function AllArticles() {
  const allArticles = useSelector((state) => state.blog.blogs);

  return (
    <section className="all-sc">
      <div className="container">
        <div className="popular-content">
          <div className="section-head flex items-end justify-between">
            <div className="sc-head-l">
              <h4 className="title title-lg">All Articles</h4>
              <p className="text text-base">
                We share common trends, strategies ideas, opinions, short & long stories from the team behind company.
              </p>
            </div>
            <div className="sc-head-r">
              <Link to="/blogs/all" type="button" className="btn btn-o-primary">
                View all
                <span className="btn-icon">
                  <i className="bi bi-arrow-right" />
                </span>
              </Link>
            </div>
          </div>
          {/* --- binding top 5 posts on list- ---} */}
          <div className="card-list grid-cols grid-cols-2" data-testid="articles">
            {
                allArticles?.slice(0, 2).map((blogItem) => (
                  <Card blogItemData={blogItem} data-testid="article" showContent key={blogItem.id} />
                ))
              }
          </div>
          <div className="card-list grid-cols grid-cols-3">
            {
                allArticles?.slice(2, 5).map((blogItem) => (
                  <Card blogItemData={blogItem} data-testid="article" isCardSm key={blogItem.id} />
                ))
              }
          </div>
        </div>
      </div>
    </section>
  );
}
