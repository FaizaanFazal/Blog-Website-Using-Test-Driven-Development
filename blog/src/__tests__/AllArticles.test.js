import { screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import AllArticles from '../components/AllArticles/AllArticles';
import Card from '../components/RecentArticles/Card';
import { renderWithProviders } from '../utils/wrappertesting';
import { small } from '../utils/helperfunctions';

describe('Unit tests for AllArticles Components', () => {
  it('Snapshot test for All Article component', () => {
    const { asFragment } = renderWithProviders(<MemoryRouter><AllArticles /></MemoryRouter>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Number of Articles should be 3', () => {
    renderWithProviders(<MemoryRouter><AllArticles /></MemoryRouter>);
    const tagElements = screen.getAllByTestId('artice');
    expect(tagElements.length).toBe(5);
  });

  it('View All Button', () => {
    renderWithProviders(<MemoryRouter><AllArticles /></MemoryRouter>);
    const searchbtn = screen.getByText('View all');
    expect(searchbtn).toBeInTheDocument();
  });
// we are using same card component of recentArticle in allArticles
// Below unit tests of Card component are already in recentArticles.test.js that will be separated later when i merge it into dev branch
//   'Image renders in card component' 
//   'Card link has right slug'
//   'Image has ALT'
});
