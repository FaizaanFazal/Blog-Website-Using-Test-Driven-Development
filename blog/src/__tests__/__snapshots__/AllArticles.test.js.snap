// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for AllArticles Components Snapshot test for All Article component 1`] = `
<DocumentFragment>
  <div>
    AllArticles
  </div>
</DocumentFragment>
`;
